### PR TITLE
[Merged by Bors] - feat(GroupTheory/GroupExtension/Defs): define group extensions

### PR DESCRIPTION
**Pull request by yu-yama**
Mainly defines:

- `structure GroupExtension N E G`: group extensions as short exact sequences `1 → N → E → G → 1`
- `structure GroupExtension.Equiv S S'`: equivalences of two group extensions `S : GroupExtension N E G` and `S' : GroupExtension N E' G`
- `structure GroupExtension.Splitting S`: splittings of a group extension `S`
- `def SemidirectProduct.toGroupExtension (φ : G →* MulAut N)`: the group extension associated with the semidirect product, `1 → N → N ⋊[φ] G → G → 1`

---

The motivation of this PR is to add definitions required to formalise relationships between classification of group extensions by a commutative group and group cohomology, especially:
- a bijection between conjugacy classes of splittings of `SemidirectProduct.toGroupExtension φ` and `H1` (which is done in [my repository](https://github.com/yu-yama/GroupExtension/blob/main/GroupExtension/Abelian.lean#L119)); and
- a bijection between equivalence classes of group extensions and `H2` (which is WIP and also noted as a TODO here).
<https://github.com/leanprover-community/mathlib4/blob/de6c2cd20c87474b696039a707b4c6348087c40a/Mathlib/RepresentationTheory/GroupCohomology/LowDegree.lean#L47-L49>

For now, I am opening this PR with the `Defs` file only, to get your opinions on the definitions. The other files are available in [my repository](https://github.com/yu-yama/GroupExtension), and I am happy to open another PR if that would be more convenient.

I would appreciate your comments.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/GroupTheory/GroupExtension/Defs.lean [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+```text
+      ↗︎ E  ↘
+1 → N   ↓    G → 1
+      ↘︎ E' ↗︎️
+```
+
+- `GroupExtension.Splitting S`: structure for splittings of a group extension `S` of `G` by `N` as
+  section homomorphisms `G → E`
+- `SemidirectProduct.toGroupExtension φ`: a canonical group extension giving a semidirect product,
+  `1 → N → N ⋊[φ] G → G → 1`
```
Comment (kbuzzard) [inline Mathlib/GroupTheory/GroupExtension/Defs.lean:32]:
```suggestion
- `SemidirectProduct.toGroupExtension φ`: the group extension associated to the semidirect product coming from `φ : G →* MulAut N`: `1 → N → N ⋊[φ] G → G → 1`
```

File: Mathlib/GroupTheory/GroupExtension/Defs.lean [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+  /-- The section is a left inverse of the projection map -/
+  rightHom_comp_sectionHom : S.rightHom.comp sectionHom = MonoidHom.id G
+
+instance : FunLike (S.Splitting) G E where
+  coe s := s.sectionHom
+  coe_injective' := by
+    intro ⟨_, _⟩ ⟨_, _⟩ h
+    congr
+    exact DFunLike.coe_injective h
+
```
Comment (kbuzzard) [inline Mathlib/GroupTheory/GroupExtension/Defs.lean:119]: Maybe you should also make an instance of `MonoidHomClass`?

File: Mathlib/GroupTheory/GroupExtension/Defs.lean [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+  /-- The projection map is surjective -/
+  rightHom_surjective : Function.Surjective rightHom
+
+variable {N E G}
+
+namespace GroupExtension
+
+variable (S : GroupExtension N E G)
+
+/-- The range of the inclusion map is a normal subgroup -/
```
Comment (kbuzzard) [inline Mathlib/GroupTheory/GroupExtension/Defs.lean:66]: Here and in several other places the sentence should end with a full stop.

File: Mathlib/GroupTheory/GroupExtension/Defs.lean [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+    congr
+    exact DFunLike.coe_injective h
+
+instance : MonoidHomClass (S.Splitting) G E where
+  map_mul := fun s ↦ s.sectionHom.map_mul'
+  map_one := fun s ↦ s.sectionHom.map_one'
+
+/-- A splitting of an extension `S` is `N`-conjugate to another iff there exists `n : N` such that
+the section homomorphism is a conjugate of the other section homomorphism by `S.inl n`. -/
+def IsConj (S : GroupExtension N E G) (s s' : S.Splitting) :=
```
Comment (eric-wieser) [inline Mathlib/GroupTheory/GroupExtension/Defs.lean:126]:
```suggestion
def IsConj (S : GroupExtension N E G) (s s' : S.Splitting) : Prop :=
```

File: Mathlib/GroupTheory/GroupExtension/Defs.lean [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+noncomputable def conjAct : E →* MulAut N := {
+  toFun := fun e ↦ (MonoidHom.ofInjective S.inl_injective).trans <|
+    (MulAut.conjNormal e).trans (MonoidHom.ofInjective S.inl_injective).symm
+  map_one' := by
+    ext _
+    simp only [map_one, MulEquiv.trans_apply, MulAut.one_apply, MulEquiv.symm_apply_apply]
+  map_mul' := fun _ _ ↦ by
+    ext _
+    simp only [map_mul, MulEquiv.trans_apply, MulAut.mul_apply, MulEquiv.apply_symm_apply]
+}
```
Comment (eric-wieser) [inline Mathlib/GroupTheory/GroupExtension/Defs.lean:89]:
```suggestion
noncomputable def conjAct : E →* MulAut N where
  toFun e := (MonoidHom.ofInjective S.inl_injective).trans <|
    (MulAut.conjNormal e).trans (MonoidHom.ofInjective S.inl_injective).symm
  map_one' := by
    ext _
    simp only [map_one, MulEquiv.trans_apply, MulAut.one_apply, MulEquiv.symm_apply_apply]
  map_mul' _ _ := by
    ext _
    simp only [map_mul, MulEquiv.trans_apply, MulAut.mul_apply, MulEquiv.apply_symm_apply]
```

File: Mathlib/GroupTheory/GroupExtension/Defs.lean [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+  S.range_inl_eq_ker_rightHom ▸ S.rightHom.normal_ker
+
+theorem rightHom_inl (n : N) : S.rightHom (S.inl n) = 1 := by
+  rw [← MonoidHom.mem_ker, ← S.range_inl_eq_ker_rightHom, MonoidHom.mem_range]
+  exact exists_apply_eq_apply S.inl n
+
+theorem rightHom_comp_inl : S.rightHom.comp S.inl = 1 := by
+  ext n
+  rw [MonoidHom.one_apply, MonoidHom.comp_apply]
+  exact S.rightHom_inl n
```
[the remark this answers is not in the document]
Comment (yu-yama) [reply Mathlib/GroupTheory/GroupExtension/Defs.lean:77]: Thank you for the suggestions, but I don't quite get what "the RHS should perhaps be `.id`" means. Could you elaborate a little bit?

[the remark this answers is not in the document]
Comment (kbuzzard) [reply Mathlib/GroupTheory/GroupExtension/Defs.lean]: These 1's are not identity functions, they're the trivial element of G so I think this is OK.

**Comment by kbuzzard** [comment]
Thanks!

maintainer merge

**Review by riccardobrasca — COMMENTED** [review]
Do we want to additivize everything?

**Comment by yu-yama** [comment]
@riccardobrasca Thank you for your review.

For now, I have additivised everything, **except** for:

- `GroupExtension.conjAct : E →* MulAut N` (and a related lemma), because `AddGroupExtension.conjAct : E →+ AddAut N` would require `AddAut N` to be an additive group; and
- `SemidirectProduct.toGroupExtension` (and related lemmas and a definition), because it depends on `SemidirectProduct` which is only defined for multiplicative groups in Mathlib.

For the former, one workaround I came up with is the following, but I think it is going to be tedious to manually write `Multiplicative` every time we use it.

```lean4
noncomputable def AddGroupExtension.conjAct : Multiplicative E →* AddAut N where
  toFun e := (AddMonoidHom.ofInjective S.inl_injective).trans <|
    ((AddEquiv.toMultiplicative (G := S.inl.range) (H := S.inl.range)).symm
      (MulAut.conjNormal (H := AddSubgroup.toSubgroup S.inl.range) e)).trans
    (AddMonoidHom.ofInjective S.inl_injective).symm
  map_one' := by
    ext _
    simp only [map_one, AddEquiv.trans_apply, AddEquiv.toMultiplicative_symm_apply_apply,
      MulEquiv.toMonoidHom_eq_coe, AddMonoidHom.toMultiplicative_symm_apply_apply,
      MonoidHom.coe_coe, MulAut.one_apply, toAdd_ofAdd, AddEquiv.symm_apply_apply, AddAut.one_apply]
  map_mul' _ _ := by
    ext _
    simp only [map_mul, AddEquiv.trans_apply, AddEquiv.toMultiplicative_symm_apply_apply,
      MulEquiv.toMonoidHom_eq_coe, AddMonoidHom.toMultiplicative_symm_apply_apply,
      MonoidHom.coe_coe, MulAut.mul_apply, AddAut.mul_apply, AddEquiv.apply_symm_apply, ofAdd_toAdd]
```

Is there a better definition? (Or do we want to leave it not additivised?)
Any suggestions would be appreciated.

**Comment by riccardobrasca** [comment]
LGTM, thanks!

bors merge

**Pull request closed without merging**